sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,336 @@
+
+            if (methodSymbol == null ||
+                methodSymbol.Parameters.Length == 0 ||
+                methodSymbol.IsOverride)
+            {
+                return Enumerable.Empty<Diagnostic>();
+            }
+
+            var parametersToCheck = methodSymbol.Parameters
+                .Where(IsTrackedParameter)
+                .ToDictionary(p => p.Name, p => new ParameterData(p));
+
+            var parameterUsesInMethod = context
+                .Node
+                .DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .Where(id => parametersToCheck.Values.Any(p => p.MatchesIdentifier(id, context.SemanticModel)));
+
+            foreach (var identifierReference in parameterUsesInMethod)
+            {
+                var key = identifierReference.Identifier.ValueText ?? "";
+
+                ParameterData paramData;
+                if (!parametersToCheck.TryGetValue(key, out paramData) ||
+                    !paramData.ShouldReportOn)
+                {
+                    continue;
+                }
+
+                if (identifierReference.Parent is EqualsValueClauseSyntax ||
+                    identifierReference.Parent is AssignmentExpressionSyntax)
+                {
+                    paramData.ShouldReportOn = false;
+                    continue;
+                }
+
+                var symbolUsedAs = FindParameterUseAsType(identifierReference, context.SemanticModel);
+                if (symbolUsedAs != null)
+                {
+                    paramData.AddUsage(symbolUsedAs);
+                }
+            }
+
+            return parametersToCheck
+                .Values
+                .Select(p => p.GetRuleViolation())
+                .WhereNotNull();
+        }
+
+        private static bool IsTrackedParameter(IParameterSymbol parameterSymbol)
+        {
+            var type = parameterSymbol.Type;
+
+            return !type.DerivesFrom(KnownType.System_Array) &&
+                   !type.IsValueType &&
+                   !type.Is(KnownType.System_String);
+        }
+
+        private SyntaxNode GetNextUnparenthesizedParent(SyntaxNode node)
+        {
+            var expression = node as ExpressionSyntax;
+            if (expression == null)
+            {
+                return node;
+            }
+
+            var topmostParent = SyntaxHelper.GetSelfOrTopParenthesizedExpression(expression);
+            return topmostParent.Parent;
+            //return (topmostParent is ParenthesizedExpressionSyntax)
+            //    ? topmostParent.Parent
+            //    : node;
+        }
+
+        private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax identifier, SemanticModel semanticModel)
+        {
+            var identifierParent = GetNextUnparenthesizedParent(identifier);
+
+            if (identifierParent is ConditionalAccessExpressionSyntax)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this type-check-and-cast sequence with an 'as' and a null check. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3247)